accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+
+The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
+- Type
+  - Value types
+    - Built-in value types
+    - User-defined value types
+    - Enumerations
+  - Reference types
+    - **Pointer types**
+    - Interface types
+    - Self-describing types
+      - Arrays
+      - Class types
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // These throw InvalidOperationException if IsFunctionPointer is false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
+        // Required since this can't be determined by inspecting the results from GetFunctionPointerCallingConventions()
++       public virtual bool IsUnmanagedFunctionPointer { get; }
+
+        // Exposing here without requiring a signature type is debatable; it is here since this
+        // information may be useful for runtime scenarios, and not just metadata inspection.
++       public virtual Type[] GetFunctionPointerCallingConventions(); 
+    }
+}
+```
+
+## System.Reflection.TypeDelegator
+No new APIs, but new overrides to support delegation.
+```diff
+namespace System.Reflection
+{
+    public class TypeDelegator : TypeInfo
+    {
++       public override Type GetFunctionPointerReturnType();
++       public override Type[] GetFunctionPointerParameterTypes();
++       public override bool IsFunctionPointer { get; }
++       public override bool IsUnmanagedFunctionPointer { get; }
++       public override Type[] GetFunctionPointerCallingConventions(); 
+    }
+}
+```
+
+## Signature types in MetadataLoadContext assembly
+These extensions are located in the `MetadataLoadContext` assembly only.
+
+Instead of extensions we could add directly to the appropriate types (`Type` \ `FieldInfo` \ `PropertyInfo` \ `ParameterInfo`) as a hedge if we think we will want the runtime to implement these in the future. They would just throw an exception if used in the runtime for now. If we don't do that, and use `MetadataLoadContext` extensions as proposed, but then attempt to add support to the runtime later that will require a new set of extensions in the runtime assembly which may cause compiler-time ambiguity as to which extensions will be called.
+
+```diff
+namespace System.Reflection
+{
++   public static class MetadataLoadContextExtensions
++   {
+        // These support function pointer types on FieldInfo.FieldType, PropertyInfo.PropertyType and ParameterInfo.ParameterType.
+        // InvalidOperationException is thrown when the corresponding type.IsFunctionPointer==false.
++       public static Type GetSignatureType(this FieldInfo fieldInfo);
```

Comment:
These names should match existing non-signature names. We have `FieldInfo.FieldType`, so the signature type getter should be `FieldInfo.GetSignatureFieldType()`

Reply:
Yep thanks